cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/io/cdap/cdap/etl/common/plugin/WrappedReduceAggregator.java

```diff
@@ -0,0 +1,122 @@
+ * are setup correctly.
+ *
+ * @param <GROUP_KEY> group key type. Must be a supported type
+ * @param <GROUP_VALUE> group value type. Must be a supported type
+ * @param <OUT> output object type
+ */
+public class WrappedReduceAggregator<GROUP_KEY, GROUP_VALUE, OUT>
+  extends BatchReduceAggregator<GROUP_KEY, GROUP_VALUE, OUT>  {
+  private final BatchReduceAggregator<GROUP_KEY, GROUP_VALUE, OUT> aggregator;
+  private final Caller caller;
+  private final OperationTimer operationTimer;
+
+  public WrappedReduceAggregator(BatchReduceAggregator<GROUP_KEY, GROUP_VALUE, OUT> aggregator, Caller caller,
+                                 OperationTimer operationTimer) {
+    this.aggregator = aggregator;
+    this.caller = caller;
+    this.operationTimer = operationTimer;
+  }
+
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+    caller.callUnchecked((Callable<Void>) () -> {
+      aggregator.configurePipeline(pipelineConfigurer);
+      return null;
+    });
+  }
+
+  @Override
+  public void initialize(BatchRuntimeContext context) throws Exception {
+    caller.call((Callable<Void>) () -> {
+      aggregator.initialize(context);
+      return null;
+    });
+  }
+
+  @Override
+  public void destroy() {
+    caller.callUnchecked((Callable<Void>) () -> {
+      aggregator.destroy();
+      return null;
+    });
+  }
+
+  @Override
+  public void prepareRun(BatchAggregatorContext context) throws Exception {
+    context.setGroupKeyClass(TypeChecker.getGroupKeyClass(aggregator));
+    context.setGroupValueClass(TypeChecker.getGroupValueClass(aggregator));
+    caller.call((Callable<Void>) () -> {
+      aggregator.prepareRun(context);
+      return null;
+    });
+  }
+
+  @Override
+  public void onRunFinish(boolean succeeded, BatchAggregatorContext context) {
+    caller.callUnchecked((Callable<Void>) () -> {
+      aggregator.onRunFinish(succeeded, context);
+      return null;
+    });
+  }
+
+  @Override
+  public void groupBy(GROUP_VALUE groupValue, Emitter<GROUP_KEY> emitter) throws Exception {
+    operationTimer.start();
+    try {
+      caller.call((Callable<Void>) () -> {
+        aggregator.groupBy(groupValue, new UntimedEmitter<>(emitter, operationTimer));
+        return null;
+      });
+    } finally {
+      operationTimer.reset();
+    }
+  }
+
+  @Override
+  public void aggregate(GROUP_KEY groupKey, GROUP_VALUE groupValues,
+                        Emitter<OUT> emitter) throws Exception {
+    operationTimer.start();
+    try {
+      caller.call((Callable<Void>) () -> {
+        aggregator.aggregate(groupKey, groupValues, new UntimedEmitter<>(emitter, operationTimer));
+        return null;
+      });
+    } finally {
+      operationTimer.reset();
+    }
+  }
+
+  @Override
+  public GROUP_VALUE reduce(GROUP_VALUE value1, GROUP_VALUE value2) throws Exception {
+    return caller.call(() -> aggregator.reduce(value1, value2));
```

Comment:
timer should be started for this too, this is where most of the work is happening.